slasher/detection/filter/filter.go

```diff
@@ -0,0 +1,94 @@
+package filter
+
+import (
+	"encoding/binary"
+	"hash/fnv"
+
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+	"github.com/spaolacci/murmur3"
+)
+
+// Filter is an encoded set of a []byte key.
+type Filter []byte
+
+// NewFilter returns a new Bloom filter that encodes the given key with 16 bits allotted for it.
+func NewFilter(key []byte) (Filter, error) {
+	nBits := 16
+	filter := make([]byte, 2)
+	for i := 0; i < len(filter); i++ {
+		filter[i] = 0
+	}
+
+	hash1 := hashutil.FastSum64(key)
+	bitPos := hash1 % uint64(nBits)
+	filter[bitPos/8] |= 1 << (bitPos % 8)
+
+	hash2 := murmur3.Sum64(key)
+	bitPos = hash2 % uint64(nBits)
+	filter[bitPos/8] |= 1 << (bitPos % 8)
+
+	hash3 := fnv.New64()
+	if _, err := hash3.Write(key); err != nil {
+		return nil, err
+	}
+	bitPos = hash3.Sum64() % uint64(nBits)
+	filter[bitPos/8] |= 1 << (bitPos % 8)
+
+	hash4 := hashutil.Hash(key)
+	hash64 := binary.LittleEndian.Uint64(hash4[:])
+	bitPos = hash64 % uint64(nBits)
+	filter[bitPos/8] |= 1 << (bitPos % 8)
+
+	hash5 := hashutil.HashKeccak256(key)
+	hash64 = binary.LittleEndian.Uint64(hash5[:])
+	bitPos = hash64 % uint64(nBits)
+	filter[bitPos/8] |= 1 << (bitPos % 8)
+
+	return filter, nil
+}
+
+// Contains returns whether the filter contains given key. False positives
+// are possible, where it could return true for a key not in the original set.
+func (f Filter) Contains(key []byte) (bool, error) {
+	if len(f) < 2 {
+		return false, nil
+	}
+	nBits := uint64(16)
+
+	hash1 := hashutil.FastSum64(key)
```

Comment:
Maybe explain the different hash functions and why you're using them (also why you're using 3 of them)